src/daemon/rpmostreed-os.cxx

```diff
@@ -932,6 +934,127 @@ os_handle_finalize_deployment (RPMOSTreeOS *interface, GDBusMethodInvocation *in
   return TRUE;
 }
 
+static DnfContext *
+os_create_bare_dnf_context (GCancellable *cancellable, GError **error)
+{
+  g_autoptr (DnfContext) dnfctx = dnf_context_new ();
+  g_autoptr (DnfState) state = NULL;
+
+  /* Use a bare DNF context for the search */
+  dnf_context_set_repo_dir (dnfctx, "/etc/yum.repos.d");
+  dnf_context_set_cache_dir (dnfctx, RPMOSTREE_CORE_CACHEDIR RPMOSTREE_DIR_CACHE_REPOMD);
+  dnf_context_set_solv_dir (dnfctx, RPMOSTREE_CORE_CACHEDIR RPMOSTREE_DIR_CACHE_SOLV);
+  dnf_context_set_cache_age (dnfctx, G_MAXUINT);
+  dnf_context_set_enable_filelists (dnfctx, FALSE);
+
+  if (!dnf_context_setup (dnfctx, cancellable, error))
+    return NULL;
+
+  state = dnf_state_new ();
+  if (!dnf_context_setup_sack_with_flags (dnfctx, state, DNF_CONTEXT_SETUP_SACK_FLAG_SKIP_RPMDB,
+                                          error))
+    return NULL;
+
+  return (DnfContext *)g_steal_pointer (&dnfctx);
+}
+
+static void
+os_add_package_info_to_builder (DnfPackage *pkg, GVariantBuilder *builder, const gchar *key)
+{
+  GVariantDict pkg_dict;
+  g_variant_dict_init (&pkg_dict, NULL);
+  if (key != NULL)
+    g_variant_dict_insert (&pkg_dict, "key", "s", key);
+#define add_value_safe(_key, _value)                                                               \
+  g_variant_dict_insert (&pkg_dict, _key, "s", _value ? _value : "");
+  add_value_safe ("name", dnf_package_get_name (pkg));
+  g_variant_dict_insert (&pkg_dict, "epoch", "t", dnf_package_get_epoch (pkg));
+  add_value_safe ("version", dnf_package_get_version (pkg));
+  add_value_safe ("arch", dnf_package_get_arch (pkg));
```

Comment:
Actually, this would likely be the 3rd "dnf package to variant" we have.  Most likely target for dedup looks like `rpmostree_variant_pkgs_from_sack()`?  We also have `single_pkg_variant_new()`.

Today also, we've gone to some effort to avoid serializing all package metadata over DBus; the CLI client for example computes rpmdb diffs on its side. 

I guess in this case there's unlikely to be *lots* of results, but if we ever implement package search we'll want to do something like serializing all the results to a memfd and passing that back to the client.  (Or streaming results)

Reply:
Aha, I wasn't aware of the similar functions. I see each of these returns a different set of properties. I added those properties the gnome-software uses. They are also named, thus it's easier to know what is what, instead of reading the arguments positionally (index 1 means `name`, index 2 `evr`,...).

I'm afraid I'm lost on the memfd side, but I agree, if there is a lot of data to be transferred around, then it makes sense. Maybe it is the case already, because I plan to ask for the package information in batch, for multiple packages at once. How many at once I do not know. My tests were less than or around 10 packages.